libass/ass.c

```diff
@@ -1359,6 +1374,52 @@ void ass_flush_events(ASS_Track *track)
     track->parser_priv->read_order_elems = 0;
 }
 
+void ass_configure_gc(ASS_Track *track, int enabled, long long delay)
+{
+    track->parser_priv->gc_enabled = enabled;
+
+    if (0ll <= delay && enabled)
+        track->parser_priv->gc_delay = delay;
+}
+
+void ass_prune_events(ASS_Track *track, long long deadline)
+{
+    if (deadline < track->parser_priv->gc_next_undisplay_ts)
+        return;
+
+    const bool check_readorder = track->parser_priv->check_readorder;
+    const int old_n_events = track->n_events;
+
+    ASS_Event *events = track->events;
+
+    long long min_undisplay_remaining = LLONG_MAX;
+    int n_kept = 0;
+
+    for (int k = 0; k < old_n_events;) {
+        // discardable sequence
+        for (; k < old_n_events && events[k].Start + events[k].Duration < deadline; k++) {
+            if (check_readorder)
+                clear_read_order_bit(track, events[k].ReadOrder);
+            ass_free_event(track, k);
+        }
+
+        // to-be-kept sequence
+        int move_from = k;
+        for (; k < old_n_events && events[k].Start + events[k].Duration >= deadline; k++)
+            ASS_CACHE_MIN_TS(min_undisplay_remaining, events[k].Start + events[k].Duration);
+
+        // Relocate kept events
+        if (move_from < k) {
+            int cnt = k - move_from;
+            memmove(events + n_kept, events + move_from, cnt * sizeof(*track->events));
+            n_kept += cnt;
+        }
+    }
```

Comment:
I think straightforward approach would be simpler to understand and not that much different in performance:
```c
    int n = 0;
    for (int i = 0; i < old_n_events; i++) {
        if (events[i].Start + events[i].Duration < deadline) {
            if (check_readorder)
                clear_read_order_bit(track, events[i].ReadOrder);
            ass_free_event(track, i);
        } else {
            ASS_CACHE_MIN_TS(min_undisplay_remaining, events[i].Start + events[i].Duration);
            events[n] = events[i];
            n++;
        }
    }
```

Also I think it's better to move `if (check_readorder)` check inside `clear_read_order_bit()`.

Reply:
> Also I think it's better to move `if (check_readorder)` check inside `clear_read_order_bit()`.

That would be an inconsistency between `test_and_set_read_order_bit` and `clear_read_order_bit`. I would suggest to not do anything, or inline both with the check moved inside.

Reply:
`memmove` can be faster than simple loop, but in this case it's simple scan + `memmove`. I think one simple loop would be faster, especially in the case where continuous bunch of events is larger than L1 cache. But yes, it should be benchmarked.

Reply:
I did some very basic artificial benchmarking with `__rdtsc()`, 50,000 runs each. No particular gcc optimization.

Given a list of 1000 events:
1. Element-wise interleaving: every other event is to be pruned.
```
    memmove: 11330.00888   1-1copy: 6737.40456
```
2. Block-wise interleaving: every other chunk of 25 events is to be pruned:<br/>
```
    memmove: 5906.99984    1-1copy: 6829.36928
```
3. Ordered list: only one event the middle is to be pruned.
```
    memmove: 4810.76426    1-1copy: 8892.79756
```

Reply:
> No particular gcc optimization.

Do you mean the default `-O2`? There's no sense in benchmarking `-O0` builtds.

Anyway, I suspect that it's not due to `memmove`, but due to `if (move_from < k)` check. Can you test one event at the start, instead of the middle? Similar optimization can be added to direct copy approach too, but given that discarded events usually would be at the start, I think it's not worth it.

Reply:
`-O0` had been used, but the tendency is similar with `-O2`.
Only the first event is discarded:
```
    memmove: 1342.11    1-1copy: 2912.24
```

With `-O2`, equivalence is at a chunk size of 7-8.